### DISCUSSION TITLE
Availability machine.USBDevice on stm32

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#16060] [STM32 / Pyboard]
Are there plans to make the USBDevice class available on the stm32 platforms?
If needed, I would like to contribute but I am not a specialist in USB stuff. Is it easy to port from the rp2040 code to the stm32?

**Comment by andrewleech** [comment]
The machine.USBDevice API is built on top of the TinyUSB library, whereas the stm32 port currently uses stm's native USB stack. 

I've been working towards migrating stm32 over to TinyUSB in https://github.com/micropython/micropython/pull/15592 in my personal time, once that's finished the runtime USB support should just work.

Before I can finish that though, the stm32 port has some other more advanced USB features like the CDC / VCP interface being exposed to python with the default repl being configured via dupterm. This provides the ability at runtime to disconnect the USB CDC from repl and use it for custom data. 
I'm working towards implementing this functionality for all TinyUSB based ports in https://github.com/micropython/micropython/pull/16017

Once that's in I can finish migrating stm32 to TinyUSB without losing the existing features.

**Reply by saraverbeecke** [reply]
Wow @andrewleech that is good to hear. Looking forward to the merge. Any estimate timeline for this to get merged? Would it be helpful if I tested your code on a F4 or H7?

**Reply by andrewleech** [reply]
Don't have a timeline, my time for this is rather sporadic. 

The USB class stuff is being a bit trickier then expected to get working properly, there's internal shortcuts in dupterm that improve performance for built in objects like USB or UART which is currently causing data errors. This means there's a bug in my implementation which is proving tricky to isolate. I say this just as a way of explaining how this kind of development is hard to estimate, you often need to just keep testing and fixing one thing after another until it all works.

The stm TinyUSB migration will need testing on as many chips as possible. I've got a few but certainly not all, so it'll be great to get help from others here. If you subscribe to updates on the PR I'll post status updates there when it's ready for testing.

**Reply by saraverbeecke** [reply]
@andrewleech i did some tests with you tinyusb branch on a stm32h743. I used the USBDevice for MIDI over usb and it works fine.
Do you have any idea how i can enable MSC also so I can have a file drive over USB?

**Reply by andrewleech** [reply]
I haven't looked at enabling the TinyUSB MSC at all, I really hate it because it corrupts the filesystem so easily. At the moment it's far safer to use mpremote to copy file to and from a device. 
I'm also working on MTP file transfer support which should be much safer once finished and PTP has already been worked on here which might already work https://github.com/micropython/micropython/issues/8426#issuecomment-2851520725

**Reply by saraverbeecke** [reply]
Thanks for you answer that is promising indeed. Altough I use the built-in MSC support for a small part of the flash so users can upload their settings easily to the device. The device reads the settings from the FAT partition and handles is further in the other flash part using littlefs. I would like to to this as well with you branch but could not see where to enable the MSC.

**Reply by andrewleech** [reply]
Do a global search for `MICROPY_HW_USB_MSC` I think it might be optionally enabled on some other ports?
Setting that to 1 enables the TinyUSB MSC driver, however I presume extra support is needed to bridge it through to the desired filesystem on any particular port which I haven't looked into for stm32.

**Reply by saraverbeecke** [reply]
ok I'll will look for that. Bridging it to the desired filesystem is also a mystery to me, but I'll give it a try. Any idea if USB Host is possible when using TinyUSB?

**Reply by andrewleech** [reply]
I'm actually working on usbhost support for tinyusb! 
I have it working "pretty well" exposing usbhost support for hid, CDC/serial and mass storage.
I've got a bunch of multi-tests to cover both USB host and USB device (two devices plugged together) but haven't actually run the tests / finished debugging the implementation, been trying to get a suitable hardware setup in place to run all that reliably.

**Reply by saraverbeecke** [reply]
Again very nice @andrewleech! I will test this as well on my stm32 boards here and report back. I hope to use it to implement USB MIDI (host + device) (and maybe enable MSC) with this. Do you think your CDC implementation is a good starting point for the MIDI protocol?

**Reply by andrewleech** [reply]
The usbhost support isn't in a PR yet, needs a lot of cleaning up first. Currently it's just exposing the host drivers built into tinyusb; I don't know if they have host support for midi built in. Once this initial support is finished and working it should be possible to extend it to provide a runtime / python definition interface for custom drivers like USBDevice has but I'm not sure how much work that'll be.

**Reply by saraverbeecke** [reply]
@andrewleech what do I need to add in my mpconfigboard.h if I want to test your tinyusb implementation on my stm32h7 (REPL over USB and MSC support for example?) 
I added:
```
#define MICROPY_HW_TINYUSB_STACK 1
#define MICROPY_HW_ENABLE_USBDEV 1
#define MICROPY_HW_USB_CDC 1
```

but I cannot see the USB device in my devices list after a successful build and deployment.

**Reply by andrewleech** [reply]
Is this with the current version of my PR checked out?
I would have expected just `#define MICROPY_HW_TINYUSB_STACK 1` should be all it takes.  
That should get USB CDC for repl; mass storage isn't supported currently.

Maybe there's an issue with the h7, I'll have to see if I can find a high speed USB device to test with.

**Reply by saraverbeecke** [reply]
No need to test anymore, I forgot to add first `#define MICROPY_HW_ENABLE_USB 1`. I thought the `#define MICROPY_HW_TINYUSB_STACK 1` did this by default.
So I have REPL over USB working now with:
```
#define MICROPY_HW_ENABLE_USB 1
#define MICROPY_HW_TINYUSB_STACK 1
```
(yes, tested with your version of the PR checked out.)
Is it straight forward to add MSC here?

**Reply by andrewleech** [reply]
I don't know how much effort MSC would be, I hate it on embedded devices because it pretty much always results in corrupted flash.
I know it can be enabled in TinyUSB but then it'd need binding code to get from that to the raw flash block layer. 
I and a couple of others have been slowly working on a USB MTP driver which is the much safer replacement for MSC, hopefully that can be finished off soon.

**Comment by saraverbeecke** [comment]
@andrewleech I am back at this project now and am trying to get the TinyUSB stack working on a stm32h743. I am glad your fantastic work got merged in the official MP repo! Using your fork with the old stm32_tinyusb branch worked for me before the merge but using the same build settings on the official repo does not compile:
```
#define MICROPY_HW_ENABLE_USB (1)
#define MICROPY_HW_TINYUSB_STACK (1)
```
gives me: **error: #error Unable to determine proper MICROPY_HW_USB_MAIN_DEV to use**

to solve this I add:
```
#define MICROPY_HW_USB_FS (1)
```

but this give me:
```
In file included from ../../../ports/stm32/usbd_conf.c:35:
../../../shared/tinyusb/mp_usbd.h:59:10: fatal error: tusb.h: No such file or directory
   59 | #include "tusb.h"
      |          ^~~~~~~~
compilation terminated.
```

any idea what I do wrong? (i did a fresh repo clone, built the submodules recursively etc. but no luck)

**Reply by andrewleech** [reply]
Hi @saraverbeecke that chip has two USB ports doesn't it? One capable of HS (with external phy) and the other regular FS?

I did have some support for these in an earlier version of the branch but it was messy and unreliable so got pulled out into a separate add-on branch which I haven't finished yet... https://github.com/andrewleech/micropython/tree/stm32_tinyusb_hs_clean or https://github.com/andrewleech/micropython/tree/stm32_tinyusb_hs

**Reply by saraverbeecke** [reply]
Yes that is true it has two USB interfaces on the chip... I only need FS so i thought this was already ok with the current merge. Are you planning on doing you separate brance as a pull request in the MP repo when finished? Or is it not on the roadmap for MP you think?
